src/instructions/segmentation.rs

```diff
@@ -1,210 +1,278 @@
 
-use crate::structures::gdt::SegmentSelector;
+#[cfg(docsrs)]
+use crate::{registers::control::Cr4Flags, structures::gdt::GlobalDescriptorTable};
+use crate::{
+    registers::model_specific::{FsBase, GsBase, Msr},
+    structures::gdt::SegmentSelector,
+    VirtAddr,
+};
 
-/// Reload code segment register.
+/// An x86 segment
 ///
-/// Note this is special since we can not directly move
-/// to cs. Instead we push the new segment selector
-/// and return value on the stack and use retf
-/// to reload cs and continue at 1:.
-///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid code segment descriptor.
-#[inline]
-pub unsafe fn set_cs(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!(
-        "push {sel}",
-        "lea {tmp}, [1f + rip]",
-        "push {tmp}",
-        "retfq",
-        "1:",
-        sel = in(reg) u64::from(sel.0),
-        tmp = lateout(reg) _,
-        options(preserves_flags),
-    );
-
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_set_cs(u64::from(sel.0));
-}
-
-/// Reload stack segment register.
-///
-/// ## Safety
+/// Segment registers on x86 are 16-bit [`SegmentSelector`]s, which index into
+/// the [`GlobalDescriptorTable`]. The corresponding GDT entry is used to
+/// configure the segment itself. Note that most segmentation functionality is
+/// disabled in 64-bit mode. See the individual segments for more information.
+pub trait Segment {
+    /// Returns the current value of the segment register.
+    fn get_reg() -> SegmentSelector;
+    /// Reload the segment register. Depending on the segment, this may also
+    /// reconfigure the corresponding segment.
+    ///
+    /// ## Safety
+    ///
+    /// This function is unsafe because the caller must ensure that `sel`
+    /// is a valid segment descriptor, and that reconfiguring the segment will
+    /// not cause undefined behavior.
+    unsafe fn set_reg(sel: SegmentSelector);
+}
+
+/// An x86 segment which is actually used in 64-bit mode
 ///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid stack segment descriptor.
-#[inline]
-pub unsafe fn load_ss(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov ss, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+/// While most segments are unused in 64-bit mode, the FS and GS segments are
+/// still partially used. Only the 64-bit segment base address is used, and this
+/// address can be set via the GDT, or by using the `FSGSBASE` instructions.
+pub trait Segment64: Segment {
+    /// MSR containing the segment base. This MSR can be used to set the base
+    /// when [`CR4.FSGSBASE`][Cr4Flags::FSGSBASE] is not set.
+    const BASE: Msr;
+    /// Reads the segment base address
+    ///
+    /// ## Safety
+    ///
+    /// If [`CR4.FSGSBASE`][Cr4Flags::FSGSBASE] is not set, this instruction will throw a `#UD`.
```

Comment:
We typically define potential exceptions as safe in this crate because expections don't cause any memory safety problems. So mentioning this is of course a good idea, but maybe we can make this function safe.

Reply:
That's a really good point. Fixed in https://github.com/rust-osdev/x86_64/pull/258/commits/22aeec3edc250ee6c75bddd594f52350f2ef5280

I added an `Exceptions` section (to both `read_base` and `write_base`) and marked `read_base` as safe. We should probably review the rest of the crate to make sure we are being consistent here, but that's for another PR.